x/multisig/keeper/grpc_query.go

```diff
@@ -65,23 +67,76 @@ func (q Querier) Key(c context.Context, req *types.KeyRequest) (*types.KeyRespon
-		return types.KeyResponse_Participant{
-			Weight: key.GetWeight(p),
-			PubKey: funcs.MustOk(key.GetPubKey(p)).String(),
+	participants := slices.Map(key.GetParticipants(), func(p sdk.ValAddress) types.KeygenParticipant {
+		return types.KeygenParticipant{
+			Address: p.String(),
+			Weight:  key.GetWeight(p),
+			PubKey:  funcs.MustOk(key.GetPubKey(p)).String(),
 		}
 	})
 	sort.SliceStable(participants, func(i, j int) bool {
 		return participants[i].Weight.GT(participants[j].Weight)
 	})
 
 	return &types.KeyResponse{
-		KeyID:           req.KeyID,
-		State:           key.GetState(),
-		Height:          key.GetHeight(),
-		Timestamp:       key.GetTimestamp(),
-		ThresholdWeight: key.GetMinPassingWeight(),
-		BondedWeight:    key.GetBondedWeight(),
-		Participants:    participants,
+		KeyID:              req.KeyID,
+		State:              key.GetState(),
+		StartedAt:          key.GetHeight(),
+		StartedAtTimestamp: key.GetTimestamp(),
+		ThresholdWeight:    key.GetMinPassingWeight(),
+		BondedWeight:       key.GetBondedWeight(),
+		Participants:       participants,
+	}, nil
+}
+
+// KeygenSession returns the keygen session info for the given key ID
+func (q Querier) KeygenSession(c context.Context, req *types.KeygenSessionRequest) (*types.KeygenSessionResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+
+	session, ok := q.keeper.GetKeygenSession(ctx, req.KeyID)
+
+	var key exported.Key
+	var state exported.MultisigState
+	if !ok {
+		key, ok = q.keeper.GetKey(ctx, req.KeyID)
+		if !ok {
+			return nil, status.Error(codes.NotFound, sdkerrors.Wrap(types.ErrMultisig, fmt.Sprintf("key not found for key id [%s]", req.KeyID)).Error())
+		}
+
+		session = types.KeygenSession{KeygenThreshold: utils.ZeroThreshold}
+		state = exported.Completed
+	} else {
+		key = &session.Key
+		state = session.GetState()
+	}
+
+	snapshot := key.GetSnapshot()
+	participants := slices.Map(snapshot.GetParticipantAddresses(), func(p sdk.ValAddress) types.KeygenParticipant {
+		var pubKey string
+		if pub, ok := key.GetPubKey(p); ok {
+			pubKey = pub.String()
+		}
+
+		return types.KeygenParticipant{
+			Address: p.String(),
+			Weight:  snapshot.GetParticipantWeight(p),
+			PubKey:  pubKey,
+		}
+	})
+	sort.SliceStable(participants, func(i, j int) bool {
+		return participants[i].Weight.GT(participants[j].Weight)
+	})
+
+	return &types.KeygenSessionResponse{
+		StartedAt:              key.GetHeight(),
+		StartedAtTimestamp:     key.GetTimestamp(),
+		ExpiresAt:              session.GetExpiresAt(),
+		CompletedAt:            session.GetCompletedAt(),
+		GracePeriod:            session.GetGracePeriod(),
+		State:                  state,
+		KeygenThresholdWeight:  key.GetSnapshot().CalculateMinPassingWeight(session.KeygenThreshold),
+		SigningThresholdWeight: key.GetMinPassingWeight(),
+		BondedWeight:           key.GetBondedWeight(),
+		Participants:           participants,
```

Comment:
I find this response very weird, half of the fields are nonsensical depending on the state, eg if the session was already deleted many fields are zero values. In which context do we need this query? Does it make sense to break it up or at least use `one_of` return types? This smells very much like it's trying to do too much

Reply:
This started out by being a query to get just the candidates for a key, but we also want to get this info for a keygen session. Separate queries would cause a lot of duplication and depending on when you query it, the session might have been deleted already, so we decided to consolidate it.

How about I put the session metadata into a separate field, which can be nil if the session was deleted?